
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.18">
    
    
      
        <title>Part 2: Genome assembly - BIO726P</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7e37652d.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#part-2-genome-assembly" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="BIO726P" class="md-header__button md-logo" aria-label="BIO726P" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            BIO726P
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Part 2: Genome assembly
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  
  Home

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="BIO726P" class="md-nav__button md-logo" aria-label="BIO726P" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    BIO726P
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Practicals for the 2025 Genome Bioinformatics module
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ssh/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SSH &amp; Your Virtual Computer!
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Apocrita_Access/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Accessing Apocrita
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="part-2-genome-assembly"><strong>Part 2: Genome assembly</strong></h1>
<hr />
<p>You need to have gone through <a href="../pt-1-read-cleaning/">Part 1: Read cleaning</a>
before starting this practical.</p>
<hr />
<h2 id="1-a-brief-assembly-example"><strong>1. A brief assembly example</strong></h2>
<p>Many different pieces of software exist for genome assembly but as mentioned in the earlier lectures we will be using <a href="https://github.com/ablab/spades"><em>SPAdes</em></a>. </p>
<p>Again for this practical you will be using the remote AWS computers created for you so please ssh back into these! </p>
<div class="admonition task">
<p class="admonition-title">Task</p>
<p>Following the same procedure from the first read cleaning practical
  <a href="../pt-1-read-cleaning/">Part 1: Read cleaning</a>, create a new main directory for today's practical (e.g., <code>2025-09-23-assembly</code>), the <code>input</code>, <code>tmp</code>, and <code>results</code> subdirectories, and the file <code>WHATIDID.txt</code> to log your commands. </p>
<p>To help get you started 
  <div class="highlight"><pre><span></span><code>mkdir 2025-09-23-assembly
</code></pre></div></p>
</div>
<p>Your directory hierarchy should look like the following</p>
<div class="admonition terminal">
<p class="admonition-title">Terminal</p>
<div class="highlight"><pre><span></span><code>2025-09-23-assembly
├── input
├── tmp
├── results
└── WHATIDID.txt
</code></pre></div>
</div>
<p>Next lets link the cleaned reads from yesterdays practical into our <code>input</code> subdirectory. We will be using these clean reads to create our assembly today! </p>
<div class="admonition task">
<p class="admonition-title">Task</p>
<p>Use the following commands to achieve this! 
  <div class="highlight"><pre><span></span><code>cd ~/2025-09-23-assembly
cd input
ln -s ~/2025-09-22-read_cleaning/results/reads.pe*.clean.fq .
cd ..
</code></pre></div></p>
</div>
<div class="admonition question">
<p class="admonition-title">Question</p>
<ul>
<li>Did you note the use of <code>*</code> in the above command?</li>
<li>What does it do? (Hint: the symbol <code>*</code> is called a wildcard)</li>
</ul>
</div>
<div class="admonition task">
<p class="admonition-title">Task</p>
<p>To assemble our cleaned reads with <em>SPAdes</em>, run the following line: 
  (<em>This will take about 10 minutes - please complete the task below in this time</em>)</p>
<div class="highlight"><pre><span></span><code>spades.py<span class="w"> </span>-o<span class="w"> </span>tmp<span class="w"> </span>-1<span class="w"> </span>input/reads.pe1.clean.fq<span class="w"> </span>-2<span class="w"> </span>input/reads.pe2.clean.fq
</code></pre></div>
<p>Like any other assembler, <em>SPAdes</em> creates many files, including a 
  <code>scaffolds.fasta</code> file that is likely to be used for follow-up 
  analyses.  </p>
<p>Copy this file to your <code>results</code> directory:</p>
<div class="highlight"><pre><span></span><code>cp<span class="w"> </span>tmp/scaffolds.fasta<span class="w"> </span>results/
</code></pre></div>
<p>Take a look at the contents of this file (e.g., to see the first 10 lines, use 
  <code>head results/scaffolds.fasta</code>, or <code>tail results/scaffolds.fasta</code> to see the
  last 10 lines).</p>
</div>
<div class="admonition question">
<p class="admonition-title">Question</p>
<p>Does it contain a lot of <strong>NNNN</strong> sequences? What do you think might be the reason for that? (Do not worry if your assembly does not contain any <strong>NNNN</strong> sequence.</p>
</div>
<div class="admonition task">
<p class="admonition-title">Task</p>
<p>While waiting for spades to run please move on to the <a href="../../Apocrita_Access/">Accessing Apocrita practical</a>. </p>
</div>
<p>To get an idea of some challenges associated to <strong>_de novo_genome assembly</strong> and the approaches used to overcome them from the following papers:</p>
<ul>
<li><a href="https://www.nature.com/articles/s41586-021-03451-0">Towards complete and error-free genome assemblies of all vertebrate species. (Rhie et al 2021)</a></li>
<li><a href="https://www.sciencedirect.com/science/article/pii/S2001037019303277">Long walk to genomics: History and current approaches to genome sequencing and assembly. (Giani et al 2020)</a>.</li>
<li><a href="https://www.nature.com/articles/nmeth.3865">A hybrid approach for <em>de novo</em> human genome sequence assembly and phasing. Mostovoy et al (2016)</a>.</li>
<li><a href="https://www.nature.com/articles/nrg3933">Genetic variation and the <em>de novo</em> assembly of human genomes. Chaisson et al 2015 NRG</a></li>
</ul>
<hr />
<h2 id="2-quality-assessment"><strong>2. Quality assessment</strong></h2>
<p><em>How do we know if our genome is good?</em></p>
<blockquote>
<p>"... the performance of different <em>de novo</em> genome assembly algorithms can 
vary greatly on the same dataset, although it has been repeatedly demonstrated
that no single assembler is optimal in every possible quality metric 
[6, 7, 8]. The most widely used metrics for evaluating an assembly include 1)
contiguity statistics such as scaffold and contig N50 size, 2) accuracy
statistics such as the number of structural errors found when compared with an
available reference genome (GAGE (Genome Assembly Gold Standard Evaluation)
evaluation tool [8]), 3) presence of core eukaryotic genes (CEGMA (Core 
Eukaryotic Genes Mapping Approach) [9]) or, if available, transcript mapping
rates, and 4) the concordance of the sequence with remapped paired-end and
mate-pair reads (REAPR (Recognizing Errors in Assemblies using Paired Reads) 
[10], assembly validation [11], or assembly likelihood [12])." -  <a href="http://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0764-4">Wences &amp; Schatz (2015)</a></p>
</blockquote>
<hr />
<h3 id="21-simple-metrics"><strong>2.1 Simple metrics</strong></h3>
<p>An assembly software will generally provide some statistics about what it did.
But, note that the output formats may differ between assemblers. 
<a href="http://quast.sourceforge.net/quast"><em>Quast</em></a>,
the <em>Quality Assessment Tool for Genome Assemblies</em> is a tool designed to
generate a standardized report.</p>
<div class="admonition task">
<p class="admonition-title">Task</p>
<p>Run <em>Quast</em> on the <code>scaffolds.fasta</code>
  file without special options to get the basic statistics:</p>
<div class="highlight"><pre><span></span><code>cd ~/2025-09-23-assembly/results
quast.py scaffolds.fasta
</code></pre></div>
<p>Have a look at the report (pdf or html) generated by <em>Quast</em> (copy the Quast's
  output directory to <code>~/www/tmp</code> and access through your browser).</p>
<div class="highlight"><pre><span></span><code>cp quast_results/latest/report.html ~/www/tmp
</code></pre></div>
<p><img alt="QUAST Output" src="../../img/QUAST_Report.png" /></p>
</div>
<div class="admonition question">
<p class="admonition-title">Question</p>
<ul>
<li>What do the values in the table mean?</li>
<li>For which values is higher better, and for which ones is smaller better?</li>
</ul>
</div>
<p>In some cases, we have prior knowledge about the expected percentage of <strong>GC</strong> 
content, the number of chromosomes, and the total genome size. This information
can be compared to the statistics present in Quast's report.</p>
<div class="admonition task">
<p class="admonition-title">Task</p>
<p>See if you can find the expected GC content of the complete Fire Ant genome? Does the GC content of the assembly match what is expected? </p>
</div>
<hr />
<h3 id="22-biologically-meaningful-measures"><strong>2.2 Biologically meaningful measures</strong></h3>
<p>Unfortunately, with many of the simple metrics, it is difficult to determine
whether the assembler did things correctly, or just haphazardly stuck lots of
reads together.</p>
<p>We often have other prior information about what to expect in this genome.
For example:</p>
<ol>
<li>if we have a reference assembly from a not-too-distant relative, we can
    expect that large genome parts will be organised in the same order, i.e., 
    <a href="https://en.wikipedia.org/wiki/Synteny"><em>synteny</em></a>.</li>
<li>If we independently created a transcriptome assembly, we can expect that
    the exons making each transcript will be mapped sequentially onto the 
    genome (see <a href="http://github.com/ksanao/TGNet">TGNet</a> for an implementation).</li>
<li>We can expect different patterns of gene content and structure between
    eukaryotes and prokaryotes.</li>
<li>Pushing this idea further, we can expect a genome to contain a single copy
    of each of the "house-keeping" genes found in related species. This is 
    applied in <em>BUSCO</em> (Benchmarking Universal Single-Copy Orthologs).<br />
    Note that:<ul>
<li><em>BUSCO</em> is a refined, modernized implementation of <a href="http://korflab.ucdavis.edu/Datasets/cegma/"><em>CEGMA</em></a> 
  (Core Eukaryotic Genes Mapping Approach). <em>CEGMA</em> examines a eukaryotic
  genome assembly for the presence and completeness of 248 "core eukaryotic genes".</li>
<li><em>Quast</em> also includes a "quick and dirty" method of finding genes.</li>
</ul>
</li>
<li>We can expect that the different scaffolds in the genome have a unimodal
      distribution in sequence read coverage. Similarly, we can expect that the
      percentage of <strong>GC</strong> content will be unimodally distributed among scaffolds. Using this idea, the <a href="https://github.com/sujaikumar/assemblage">Blobology</a> approach determined that evidence of foreign sequences in Tardigrades is largely due to extensive contamination rather than extensive horizontal gene transfer <a href="http://www.pnas.org/content/113/18/5053">Koutsovoulos et al 2016</a>.</li>
</ol>
<p><strong>It is important to understand the concepts underlying these different 
approaches.</strong></p>
<hr />
<h2 id="3-in-your-own-time"><strong>3. In your own time</strong></h2>
<p>Try to figure out what are the <a href="https://www.nature.com/articles/nrg3367">tradeoffs between <strong>de bruijn graph</strong> and 
<strong>overlap-layout-consensus</strong> assembly approaches</a>.</p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["content.tooltips", "navigation.instant", "navigation.tabs", "toc.integrate", "content.tabs.link"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.92b07e13.min.js"></script>
      
    
  </body>
</html>